highgui: wayland: Exception when showing too small width image

### System Information

OpenCV version: 4.x ( ba65d2eb0d83e6c9567a2534f8994cd4484a1635 )
Operating System / Platform: Ubuntu 24.04
Compiler & compiler version: GCC 13.2

### Detailed description

When imshow() with small width image, exception is happen.
Because window view width is too small enogh to show title bar.
(In this case, roi is used to fill area. but it has negative position).

```bash
kmtr@kmtr-VMware-Virtual-Platform:~/work/build4-main$ ./bin/opencv_test_highgui --gtest_filter="*small*"
:
: 
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from Highgui_GUI
[ RUN      ] Highgui_GUI.small_width_image
[ INFO:0@0.007] global registry.impl.hpp:114 UIBackendRegistry UI: Enabled backends(3, sorted by priority): GTK(1000); GTK3(990); GTK2(980) + BUILTIN(Wayland)
[*] DEBUG: handle_toplevel_configure: maximized=0 fullscreen=0 resizing=0 focused=0 size=[0 x 0]
Exception message: OpenCV(4.9.0-dev) /home/kmtr/work/opencv4/modules/core/src/matrix.cpp:808: error: (-215:Assertion failed) 0 <= roi.x && 0 <= roi.width && roi.x + roi.width <= m.cols && 0 <= roi.y && 0 <= roi.height && roi.y + roi.height <= m.rows in function 'Mat'

/home/kmtr/work/opencv4/modules/highgui/test/test_gui.cpp:221: Failure
Expected: waitKey(1000) doesn't throw an exception.
  Actual: it throws.
[  FAILED  ] Highgui_GUI.small_width_image (59 ms)
[----------] 1 test from Highgui_GUI (59 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (59 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] Highgui_GUI.small_width_image
```

### Steps to reproduce

```diff
diff --git a/modules/highgui/test/test_gui.cpp b/modules/highgui/test/test_gui.cpp
index 8991e8072b..7821346a99 100644
--- a/modules/highgui/test/test_gui.cpp
+++ b/modules/highgui/test/test_gui.cpp
@@ -205,6 +205,22 @@ TEST(Highgui_GUI, trackbar)
     EXPECT_NO_THROW(destroyAllWindows());
 }

+// See XXXX
+#if (!defined(ENABLE_PLUGINS) \
+      && !defined HAVE_WAYLAND )
+TEST(Highgui_GUI, DISABLED_small_width_image)
+#else
+TEST(Highgui_GUI, small_width_image)
+#endif
+{
+    const std::string window_name("trackbar_test_window");
+    cv::Mat src(1,1,CV_8UC3,cv::Scalar(0));
+    EXPECT_NO_THROW(destroyAllWindows());
+    ASSERT_NO_THROW(namedWindow(window_name));
+    ASSERT_NO_THROW(imshow(window_name, src));
+    ASSERT_NO_THROW(waitKey(1000));
+    EXPECT_NO_THROW(destroyAllWindows());
+}
```


### Issue submission checklist

- [X] I report the issue, it's not a question
- [X] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [X] I updated to the latest OpenCV version and the issue is still there
- [X] There is reproducer code and related data files (videos, images, onnx, etc)

--- a/modules/highgui/src/window_wayland.cpp
+++ b/modules/highgui/src/window_wayland.cpp
@@ -570,6 +570,9 @@ private:
     cv::Rect last_img_area_;
     bool image_changed_ = false;
 
+    int real_img_width = 0;
+    cv::Scalar const outarea_color_ = CV_RGB(0, 0, 0);
+
     void *param_ = nullptr;
     CvMouseCallback callback_ = nullptr;
 };
@@ -1535,6 +1538,28 @@ void cv_wl_viewer::set_image(cv::Mat const &image) {
         image_ = image.clone();
     }
     image_changed_ = true;
+
+    // See https://github.com/opencv/opencv/issues/25560
+    // If image_ width is too small enough to show title and buttons, expand it.
+
+    // Keep real image width to limit x position for callback functions
+    real_img_width = image_.size().width;
+
+    // Minimum width of title is not defined, so use button width * 3 instead of it.
+    const int view_min_width = cv_wl_titlebar::btn_width * 3 + cv_wl_titlebar::titlebar_min_width;
+
+    const int margin = view_min_width - real_img_width;
+    if(margin > 0)
+    {
+        copyMakeBorder(image_,               // src
+                       image_,               // dst
+                       0,                    // top
+                       0,                    // bottom
+                       0,                    // left
+                       margin,               // right
+                       cv::BORDER_CONSTANT,  // borderType
+                       outarea_color_ );     // value(color)
+    }
 }
 
 void cv_wl_viewer::set_mouse_callback(CvMouseCallback callback, void *param) {
@@ -1587,6 +1612,8 @@ void cv_wl_viewer::on_mouse(int event, cv::Point const &p, int flag) {
             int x = static_cast<int>((p.x - last_img_area_.x) * ((double) image_.size().width / last_img_area_.width));
             int y = static_cast<int>((p.y - last_img_area_.y) *
                                      ((double) image_.size().height / last_img_area_.height));
+
+            x = cv::min(x, real_img_width);
             callback_(event, x, y, flag, param_);
         }
     }

--- a/modules/highgui/test/test_gui.cpp
+++ b/modules/highgui/test/test_gui.cpp
@@ -218,7 +218,7 @@ TEST(Highgui_GUI, small_width_image)
     EXPECT_NO_THROW(destroyAllWindows());
     ASSERT_NO_THROW(namedWindow(window_name));
     ASSERT_NO_THROW(imshow(window_name, src));
-    ASSERT_NO_THROW(waitKey(1000));
+    EXPECT_NO_THROW(waitKey(10));
     EXPECT_NO_THROW(destroyAllWindows());
 }
 

--- a/modules/highgui/test/test_gui.cpp
+++ b/modules/highgui/test/test_gui.cpp
@@ -206,8 +206,7 @@ TEST(Highgui_GUI, trackbar)
 }
 
 // See https://github.com/opencv/opencv/issues/25560
-#if (!defined(ENABLE_PLUGINS) \
-      && !defined HAVE_WAYLAND )
+#if !defined(ENABLE_PLUGINS)
 TEST(Highgui_GUI, DISABLED_small_width_image)
 #else
 TEST(Highgui_GUI, small_width_image)

--- a/modules/highgui/test/test_gui.cpp
+++ b/modules/highgui/test/test_gui.cpp
@@ -205,6 +205,22 @@ TEST(Highgui_GUI, trackbar)
     EXPECT_NO_THROW(destroyAllWindows());
 }
 
+// See https://github.com/opencv/opencv/issues/25560
+#if (!defined(ENABLE_PLUGINS) \
+      && !defined HAVE_WAYLAND )
+TEST(Highgui_GUI, DISABLED_small_width_image)
+#else
+TEST(Highgui_GUI, small_width_image)
+#endif
+{
+    const std::string window_name("trackbar_test_window");
+    cv::Mat src(1,1,CV_8UC3,cv::Scalar(0));
+    EXPECT_NO_THROW(destroyAllWindows());
+    ASSERT_NO_THROW(namedWindow(window_name));
+    ASSERT_NO_THROW(imshow(window_name, src));
+    ASSERT_NO_THROW(waitKey(1000));
+    EXPECT_NO_THROW(destroyAllWindows());
+}
 
 TEST(Highgui_GUI, currentUIFramework)
 {

